zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,612 @@
+       case of restoring a backup taken while the request was in
+       flight. It is unclear whether to expect the restored backup can
+       successfully deliver mobile notifications.
+
+       TBD: Consider whether to check whether we already have a valid
+       registration before processing the deletion here.
+
+   * On other errors, including 5xx errors and network errors:
+
+     Wait in order to back off, using exponential random backoff
+     according to standard practice. Then, asynchronously, retry the
+     request.
+
+     Since the server does not have `timestamp`, it cannot check if
+     the request is already past the
+     `PUSH_REGISTRATION_LIVENESS_TIMEOUT` threshold and therefore
+     stale. It is thus reliant on the server to reject such requests.
+
+   (Forgetting the `PushDevice` record will cause the client to retry
+   the operation from scratch, if the client is still in use.
+   That way if the client keeps being used, we keep retrying and keep
+   sending the server admin error messages about it — appropriately so
+   — but we don’t accumulate ghost clients.)
+
+
+### In register-queue
+
+In the data which the server sends the client in the
+[register-queue](https://zulip.com/api/register-queue) response,
+and keeps updated via events, we add the following:
+
+* `push_account_ids`, a JSON object,
+  with one entry for each of the user's `PushDevice` records.
+
+  * The key is `PushDevice.push_account_id`.
+
+  * The value is either `"active"` or `"pending"`,
+    representing the status of the device's registration.
+
+    The status is pending if `PushDevice.bouncer_device_id` is null,
+    and active otherwise.
+
+    TODO This value is now an object.
+
+The client should look up in this map the `push_account_id` value from
+the corresponding `Account` record.  It uses the result in two ways:
+
+* This data governs whether to retry registering.  If there is no
+  matching `push_account_ids` entry, the client knows the server has
+  deleted its registration and it should re-register.
+
+* This data can be used to feed a notifications troubleshooting UI.
+
+  * If the status is active, then registration has succeeded
+    and (as far as this piece of information is concerned)
+    there is nothing to warn about.
+
+  * If the status is pending, or if there is no matching entry,
+    then registration has not succeeded
+    and push notifications will not be sent to this client.
+
+    If the `Account` record was created within the last few minutes,
+    then this may be normal. Otherwise, it represents at least a
+    temporary outage somewhere in the chain, and the user may be
+    surprised that push notifications aren't working for them. The
+    client might show a warning banner to the user in this case.
+
+  * These data structures should be appropriately linked to future
+    data structures allowing lists of clients and deleting an
+    `api_key`.
```

Comment:
I think this item doesn't add anything. It looks like it was meant as an update of this:
```diff
-TODO this doesn't really work, would need to be combined with
-  separate API keys:
-  > It can also support a UI for showing the user what devices they have
-  > connected, and revoking any they aren’t using.
```

But the goal there is that UI, which would offer a feature we don't currently have (and have long talked about). The data structures are just a means to that goal… and the primary driver among those will be actually having separate API keys in the first place. This spec doesn't really have anything to add about that.

(The text quoted with `>` is from a previous draft (#2), where I believe it was incorrectly adapted from a comment thread on a draft before that.)

Reply:
Removed in f2c1956034ca30690e83ddc3f22e1b4a13fb0656.